### PR TITLE
Add error message for empty popup #402

### PR DESCRIPTION
**Pull request by draganakovacevic**
Set different error message when creating category. One for blank input fields and other for too long description.

### CODE REVIEW

**Comment by stanislavgrujic** [comment]
Please update this branch with the latest develop.

**Comment by draganakovacevic** [comment]
Exception for too long description of error message is happening with status of 500, for others like exception with status 409,etc. or custom error message, display them.

File: web/src/main/frontend/src/base/errorHandler.js [hunk 1]
```diff
@@ -1,15 +1,15 @@
 import NotificationsStore from './stores/NotificationsStore';
 
 export const setNotificationMessage = notification => {
-  if (notification.status > 400) {
+  if (notification.status == 500) {
     NotificationsStore.setNotificationMessage({
       type: 'error',
-      message: notification.response.text
+      message: 'Category is not created. Try to reduce description'
```
Comment (stanislavgrujic) [inline web/src/main/frontend/src/base/errorHandler.js:7]: Specific messages should come from the backend, not coded in the UI. Imagine if we have 250 different error messages, and we don't have only web interface, but we have Android and iOS applications, and also expose our REST API to 3rd party service providers. That would mean that we have to maintain 750 messages (triplicated) and also that each of our clients needs to know how to add 250 message.

File: web/src/main/frontend/src/base/errorHandler.js [hunk 2]
```diff
@@ -1,15 +1,15 @@
 import NotificationsStore from './stores/NotificationsStore';
 
 export const setNotificationMessage = notification => {
-  if (notification.status > 400) {
+  if (notification.status == 500) {
```
Comment (stanislavgrujic) [inline web/src/main/frontend/src/base/errorHandler.js:4]: This is not good. You should not mapy **any**  status code to this specific message. Status code `500` is not specific to *Category not created...*, but it is a sign of an `Internal Server Error` (basically anything which is not explicitly handler by an exception handler).
Please read [this](https://www.baeldung.com/rest-api-error-handling-best-practices). You should not do anything with it, just inform yourself.

File: core/src/main/java/eu/execom/testgenerator/core/service/CategoryServiceImpl.java [hunk 3]
```diff
@@ -26,9 +26,13 @@ public CategoryServiceImpl(CategoryRepository categoryRepository, QuestionServic
...
   @Transactional
   public Category save(Category category) {
-    Category oldCategory = categoryRepository.findByName(category.getName());
+    Category oldCategory = categoryRepository.findByName(category.getName().trim());
     if (oldCategory != null) {
       throw new ValidationException("Category with this name already exists!");
+    } else if(category.getDescription().length() >= 256) {
+        throw new ValidationException("Category has too long description!");
+    } else if(category.getDescription().isEmpty() || category.getName().isEmpty())  {
+        throw new ValidationException("Category fields must not be empty! ");
```
Comment (stanislavgrujic) [inline core/src/main/java/eu/execom/testgenerator/core/service/CategoryServiceImpl.java:35]: Please validate these in the corresponding *request* classes, not in the domain level. You can use annotations such as `@Size` to limit both minimum and maximum number of characters.

File: web/src/main/frontend/src/components/forms/CategoryForm.js [hunk 4]
```diff
@@ -26,8 +27,9 @@ const hooks = {
   },
 
   onError(form) {
-    // TODO: Change generic error messages!
-    form.invalidate('This is a generic error message!');
+    return setNotificationMessage({
+      type: 'error'
```
Comment (stanislavgrujic) [inline web/src/main/frontend/src/components/forms/CategoryForm.js:31]: Is the form also invalidated with this? Please try to understand all the effects of your change and make sure that you are not breaking anything with it.

File: web/src/main/frontend/src/base/errorHandler.js [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import NotificationsStore from './stores/NotificationsStore';
 
 export const setNotificationMessage = notification => {
-  if (notification.status > 400) {
+  if (notification.status > 500) {
```
Comment (stanislavgrujic) [inline web/src/main/frontend/src/base/errorHandler.js:4]: Why do you need this? What will happen with errors that have a status code 4xx?

File: core/pom.xml [hunk 6]
```diff
@@ -13,6 +13,23 @@
...
+    <dependency>
+      <groupId>org.hibernate</groupId>
+      <artifactId>hibernate-validator</artifactId>
+      <version>5.3.5.Final</version>
+    </dependency>
+
+    <dependency>
+      <groupId>org.glassfish</groupId>
+      <artifactId>javax.el</artifactId>
+    </dependency>
```
Comment (stanislavgrujic) [inline core/pom.xml:31]: Why do we need these? Hibernate-validator is already being pulled into the code by `spring-boot-starter-web` dependency. `hibernate-validator` pulls `validation-api`. The `javax.el`, I am not sure why you need it.

File: pom.xml [hunk 7]
```diff
@@ -31,6 +31,23 @@
...
+    <dependency>
+      <groupId>org.hibernate</groupId>
+      <artifactId>hibernate-validator</artifactId>
+      <version>5.3.5.Final</version>
+    </dependency>
+
+    <dependency>
+      <groupId>org.glassfish</groupId>
+      <artifactId>javax.el</artifactId>
+    </dependency>
```
Comment (stanislavgrujic) [inline pom.xml:49]: Why do you need these here as well?

File: web/pom.xml [hunk 8]
```diff
@@ -21,6 +21,24 @@
...
+    <dependency>
+      <groupId>org.hibernate</groupId>
+      <artifactId>hibernate-validator</artifactId>
+      <version>5.3.5.Final</version>
+    </dependency>
+
+    <dependency>
+      <groupId>org.glassfish</groupId>
+      <artifactId>javax.el</artifactId>
+    </dependency>
```
Comment (stanislavgrujic) [inline web/pom.xml:40]: And why do you need it here?

File: web/src/main/frontend/src/components/forms/CategoryForm.js [hunk 9]
```diff
@@ -26,8 +27,7 @@ const hooks = {
   },
 
   onError(form) {
-    // TODO: Change generic error messages!
-    form.invalidate('This is a generic error message!');
+    form.invalidate("This is a generic error message!");
```
Comment (stanislavgrujic) [inline web/src/main/frontend/src/components/forms/CategoryForm.js:30]: You just removed the TODO, but did not change generic error messages? How is this TODO solved?

File: web/src/main/frontend/src/components/forms/CategoryForm.js [hunk 10]
```diff
@@ -2,6 +2,7 @@ import validatorjs from 'validatorjs';
 import MobxReactForm from 'mobx-react-form';
 import dvr from 'mobx-react-form/lib/validators/DVR';
 import CategoryStore from '../../base/stores/CategoryStore';
+import {setNotificationMessage} from "../../base/errorHandler";
```
Comment (stanislavgrujic) [inline web/src/main/frontend/src/components/forms/CategoryForm.js:5]: Is this needed? I don't see it being used in the code.

File: web/src/main/java/eu/execom/testgenerator/web/dto/CreateFormulaRequest.java [hunk 11]
```diff
@@ -15,9 +18,12 @@
   private String name;
 
   @NotNull
+  @Size(min = 5, max = 255, message = "Description length should be between 5 and 255 characters.")
   private String description;
 
   @NotNull
+  @Min(value = 15, message = "Test duration should not be less than 15minutes.")
+  @Max(value = 180, message = "Test duration should not be more than 180minutes.")
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/CreateFormulaRequest.java:26]: Please add space between number of minutes and word *minutes*, in both annotations.
Please phrase your messages affirmatively, it is easier to understand:
*Test duration should be at least 15 minutes.*
*Test duration should be maximum 180 minutes.*

File: web/src/main/java/eu/execom/testgenerator/web/dto/CreateQuestionRequest.java [hunk 12]
```diff
@@ -13,9 +16,12 @@
 public class CreateQuestionRequest {
 
   @NotNull
+  @Size(min = 5, max = 255, message = "Description length should be between 5 and 255 characters.")
   private String text;
 
   @NotNull
+  @Min(value = 5, message = "Points per question should not be less than 5.")
+  @Max(value = 25, message = "Points per question should not be more than 25.")
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/dto/CreateQuestionRequest.java:24]: See my comment above for affirmative messages.

**Pull request merged**